src/common/hooks/useEndorsementChain/helpers.ts

```diff
@@ -0,0 +1,166 @@
+import { Provider } from "@ethersproject/abstract-provider";
+import { TitleEscrowFactory__factory, TradeTrustToken } from "@govtechsg/token-registry/dist/contracts";
+import { Signer } from "ethers";
+import { EndorsementChain, TransferBaseEvent, TransferEvent, TransferEventType } from "../../../types";
+
+export const fetchEventTime = async (blockNumber: number, provider: Provider): Promise<number> => {
+  const eventTimestamp = (await (await provider.getBlock(blockNumber)).timestamp) * 1000;
+  return eventTimestamp;
+};
+
+export const extractEscrowAddress = async (
+  tokenRegistry: TradeTrustToken,
+  tokenId: string,
+  signer: Provider | Signer
+): Promise<string> => {
+  const titleEscrowFactoryAddress = await tokenRegistry.titleEscrowFactory();
+  const tokenRegistryAddress = await tokenRegistry.address;
+  const titleEscrowFactory = TitleEscrowFactory__factory.connect(titleEscrowFactoryAddress, signer);
+  const titleEscrowAddress = await titleEscrowFactory.getAddress(tokenRegistryAddress, tokenId);
+  return titleEscrowAddress;
+};
+
+/*
+  Get available owner/holder from list of events
+*/
+export const mergeTransferTransactions = (repeatedTransferEvents: TransferBaseEvent[]): TransferBaseEvent => {
+  let owner = "";
+  let holder = "";
+  for (const event of repeatedTransferEvents) {
+    owner = event.owner || owner;
+    holder = event.owner || holder;
+  }
+  return {
+    ...repeatedTransferEvents[0],
+    owner: owner,
+    holder: holder,
+  };
+};
+
+/*
+    Merge Transactions with the same transaction hash
+    E.g. 
+    TRANSFER_OWNERS that emits: both 
+    HOLDER_TRANSFER and OWNER_TRANSFER (on Title Escrow)
+    
+    INITIAL (Minting) that emits: 
+    (TRANSFER_OWNERS as above) and INITIAL (on Token Registry)
+    
+    SURRENDER_ACCEPTED (Shred) that emits: 
+    (TRANSFER_OWNERS as above) and SURRENDER_ACCEPTED (on Token Registry)
+*/
+export const mergeTransfers = (transferEvents: TransferBaseEvent[]): TransferBaseEvent[] => {
```

Comment:
a test will be helpful. i see that there are various for loops used, traditional for loop, `for in`, `for of` which amts to quite abit of context switching. able to refactor?

Reply:
yes pls, provide some concrete tests for documentation